subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java

```diff
@@ -121,33 +147,33 @@ private JavaToolchain resolveToolchain(JavaToolchainSpecInternal requestedSpec)
 
     private JavaToolchain query(JavaToolchainSpec spec, boolean isFallback) {
         if (spec instanceof CurrentJvmToolchainSpec) {
-            // TODO (#22023) Look into checking if this optional is present and throwing an exception
-            return asToolchain(new InstallationLocation(Jvm.current().getJavaHome(), "current JVM"), spec, isFallback).getToolchain().get();
+            return asToolchainOrThrow(new InstallationLocation(currentJvm.getJavaHome(), "current JVM"), spec, isFallback);
         }
 
         if (spec instanceof SpecificInstallationToolchainSpec) {
-            final InstallationLocation installation = new InstallationLocation(((SpecificInstallationToolchainSpec) spec).getJavaHome(), "specific installation");
-            return asToolchainOrThrow(installation, spec);
+            return asToolchainOrThrow(new InstallationLocation(((SpecificInstallationToolchainSpec) spec).getJavaHome(), "specific installation"), spec, false);
         }
 
-        Optional<JavaToolchainInstantiationResult> detectedToolchain = registry.listInstallations().stream()
-                .map(javaHome -> asToolchain(javaHome, spec, false))
-                .filter(JavaToolchainMatcher.forInstantiationResult(spec))
-                .min(JavaToolchainComparator.forInstantiationResult());
-
-        if (detectedToolchain.isPresent()) {
-            warnIfAutoProvisionedToolchainUsedWithoutRepositoryDefinitions(detectedToolchain.get());
-            return detectedToolchain.get().getToolchain().get();
-        }
-
-        InstallationLocation downloadedInstallation = downloadToolchain(spec);
-        JavaToolchain downloadedToolchain = asToolchainOrThrow(downloadedInstallation, spec);
-        registry.addInstallation(downloadedInstallation);
-        return downloadedToolchain;
+        Predicate<JvmInstallationMetadata> matcher = new JvmInstallationMetadataMatcher(spec);
```

Comment:
I split out most of the `Stream` chain, but moved the orElseGet portion into the existing `downloadToolchain` method instead of creating a new method, as the resulting combined method is still concise.